File: razer-sdk/src/test/java/com/bonepl/razersdk/animation/SimpleFrameTest.java
package com.bonepl.razersdk.animation;

import com.bonepl.razersdk.color.Color;
import com.bonepl.razersdk.color.StaticColor;
import com.bonepl.razersdk.sdk.RzKey;
import org.junit.jupiter.api.Test;

import java.util.*;

import static org.junit.jupiter.api.Assertions.*;

class SimpleFrameTest {

    @Test
    void testEmptyFrameCreation() {
        //given
        final SimpleFrame simpleFrame = new SimpleFrame();

        //when
        final Frame actualFrame = simpleFrame.getFrame();

        //then
        assertTrue(actualFrame.getKeysToColors().isEmpty());
        assertFalse(simpleFrame.hasFrame());
        assertThrows(NoSuchElementException.class, simpleFrame::getFrame);
    }

    @Test
    void testStartWithFrame() {
        //when
        final SimpleFrame simpleFrame = new SimpleFrame();

        //then
        assertTrue(simpleFrame.hasFrame());
    }

    @Test
    void testFullColor() {
        //given
        final SimpleFrame simpleFrame = new SimpleFrame(StaticColor.BLUE);

        //when
        final Map<RzKey, Color> actualKeysToColors = simpleFrame.getFrame().getKeysToColors();

        //then
        assertFalse(simpleFrame.hasFrame());
        assertEquals(RzKey.values().length, actualKeysToColors.size());
        Arrays.stream(RzKey.values()).forEach(rzKey -> assertEquals(StaticColor.BLUE, actualKeysToColors.get(rzKey)));
        assertThrows(NoSuchElementException.class, simpleFrame::getFrame);
    }

    @Test
    void testOneKey() {
        //given
        final SimpleFrame simpleFrame = new SimpleFrame(RzKey.RZKEY_1, StaticColor.RED);

        //when
        final Map<RzKey, Color> actualKeysToColors = simpleFrame.getFrame().getKeysToColors();

        //then
        assertFalse(simpleFrame.hasFrame());
        assertEquals(1, actualKeysToColors.size());
        assertEquals(StaticColor.RED, actualKeysToColors.get(RzKey.RZKEY_1));
        assertThrows(NoSuchElementException.class, simpleFrame::getFrame);
    }

    @Test
    void testMultipleKeys() {
        //given
        final Collection<RzKey> rzKeys = List.of(RzKey.RZKEY_2, RzKey.RZKEY_3);
        final SimpleFrame simpleFrame = new SimpleFrame(rzKeys, StaticColor.YELLOW);

        //when
        final Map<RzKey, Color> actualKeysToColors = simpleFrame.getFrame().getKeysToColors();

        //then
        assertFalse(simpleFrame.hasFrame());
        assertEquals(2, actualKeysToColors.size());
        rzKeys.forEach(rzKey -> assertEquals(StaticColor.YELLOW, actualKeysToColors.get(rzKey)));
        assertThrows(NoSuchElementException.class, simpleFrame::getFrame);
    }

    @Test
    void testKeyMap() {
        //given
        final EnumMap<RzKey, Color> expectedKeysToColorMap = new EnumMap<>(RzKey.class);
        expectedKeysToColorMap.put(RzKey.RZKEY_Q, StaticColor.GREEN);
        expectedKeysToColorMap.put(RzKey.RZKEY_W, StaticColor.CYAN);
        expectedKeysToColorMap.put(RzKey.RZKEY_E, StaticColor.CYAN);
        final SimpleFrame simpleFrame = new SimpleFrame(expectedKeysToColorMap);

        //when
        final Map<RzKey, Color> actualKeysToColors = simpleFrame.getFrame().getKeysToColors();

        //then
        assertFalse(simpleFrame.hasFrame());
        assertEquals(3, actualKeysToColors.size());
        assertEquals(StaticColor.GREEN, actualKeysToColors.get(RzKey.RZKEY_Q));
        assertEquals(StaticColor.CYAN, actualKeysToColors.get(RzKey.RZKEY_W));
        assertEquals(StaticColor.CYAN, actualKeysToColors.get(RzKey.RZKEY_E));
        assertThrows(NoSuchElementException.class, simpleFrame::getFrame);
    }
}

File: test/razer-sdk/animation/SimpleFrame.test.ts
import {describe, expect, test} from 'vitest';
import {SimpleFrame} from '../../../src/razer-sdk/animation/SimpleFrame.js';
import {StaticColor} from '../../../src/razer-sdk/color/StaticColor.js';
import {Color} from '../../../src/razer-sdk/color/Color.js';
import {getRzKeyValues, RzKey} from '../../../src/razer-sdk/sdk/RzKey.js';

describe('SimpleFrameTest', () => {
    test('testEmptyFrameCreation', () => {
        //given
        const simpleFrame = new SimpleFrame();

        //when
        const actualFrame = simpleFrame.pollFrame();

        //then
        expect(actualFrame.getKeysToColors().size).toBe(0);
        expect(simpleFrame.hasFrame()).toBe(false);
        expect(() => simpleFrame.pollFrame()).toThrow();
    });

    test('testStartWithFrame', () => {
        //when
        const simpleFrame = new SimpleFrame();

        //then
        expect(simpleFrame.hasFrame()).toBe(true);
    });

    test('testFullColor', () => {
        //given
        const simpleFrame = new SimpleFrame(StaticColor.BLUE);

        //when
        const actualKeysToColors = simpleFrame.pollFrame().getKeysToColors();

        //then
        expect(simpleFrame.hasFrame()).toBe(false);
        expect(actualKeysToColors.size).toBe(getRzKeyValues().length);
        for (const rzKey of getRzKeyValues()) {
            expect(actualKeysToColors.get(rzKey)).toBe(StaticColor.BLUE);
        }
        expect(() => simpleFrame.pollFrame()).toThrow();
    });

    test('testOneKey', () => {
        //given
        const simpleFrame = new SimpleFrame(RzKey.RZKEY_1, StaticColor.RED);

        //when
        const actualKeysToColors = simpleFrame.pollFrame().getKeysToColors();

        //then
        expect(simpleFrame.hasFrame()).toBe(false);
        expect(actualKeysToColors.size).toBe(1);
        expect(actualKeysToColors.get(RzKey.RZKEY_1)).toBe(StaticColor.RED);
        expect(() => simpleFrame.pollFrame()).toThrow();
    });

    test('testMultipleKeys', () => {
        //given
        const rzKeys = [RzKey.RZKEY_2, RzKey.RZKEY_3];
        const simpleFrame = new SimpleFrame(rzKeys, StaticColor.YELLOW);

        //when
        const actualKeysToColors = simpleFrame.pollFrame().getKeysToColors();

        //then
        expect(simpleFrame.hasFrame()).toBe(false);
        expect(actualKeysToColors.size).toBe(2);
        for (const rzKey of rzKeys) {
            expect(actualKeysToColors.get(rzKey)).toBe(StaticColor.YELLOW);
        }
        expect(() => simpleFrame.pollFrame()).toThrow();
    });

    test('testKeyMap', () => {
        //given
        const expectedKeysToColorMap = new Map<RzKey, Color>([
            [RzKey.RZKEY_Q, StaticColor.GREEN],
            [RzKey.RZKEY_W, StaticColor.CYAN],
            [RzKey.RZKEY_E, StaticColor.CYAN],
        ]);
        const simpleFrame = new SimpleFrame(expectedKeysToColorMap);

        //when
        const actualKeysToColors = simpleFrame.pollFrame().getKeysToColors();

        //then
        expect(simpleFrame.hasFrame()).toBe(false);
        expect(actualKeysToColors.size).toBe(3);
        expect(actualKeysToColors.get(RzKey.RZKEY_Q)).toBe(StaticColor.GREEN);
        expect(actualKeysToColors.get(RzKey.RZKEY_W)).toBe(StaticColor.CYAN);
        expect(actualKeysToColors.get(RzKey.RZKEY_E)).toBe(StaticColor.CYAN);
        expect(() => simpleFrame.pollFrame()).toThrow();
    });
});
